packages/clay-form/src/InputWithAutocomplete.tsx

```diff
@@ -0,0 +1,113 @@
+/**
+ * © 2019 Liferay, Inc. <https://liferay.com>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import ClayAutocomplete from '@clayui/autocomplete';
+import ClayDropDown from '@clayui/drop-down';
+import React, {useEffect, useRef, useState} from 'react';
+import {useKeyHandlerForList} from '@clayui/shared';
+
+interface IProps extends React.InputHTMLAttributes<HTMLInputElement> {
+	/**
+	 * Flag to indicate is loading indicator should be shown.
+	 */
+	loading?: boolean;
+
+	/**
+	 * Message displayed in dropdown when laoding is indicated.
+	 */
+	loadingMessage?: string;
+
+	/**
+	 * Items that will show up in the dropdown list
+	 */
+	items: Array<string> | Array<Object>;
+
+	/**
+	 * Function that returns the value that should be showed in the list.
+	 */
+	itemSelector?: (val: string | Object) => string;
+
+	/**
+	 * Callback for when input value changes
+	 */
+	onItemSelect: (val: string | Object) => string | Object;
+
+	/**
+	 * Callback for when input value changes
+	 */
+	onValueChange: (val: string | Object) => void;
+}
+
+export const ClayInputWithAutocomplete: React.FunctionComponent<IProps> = ({
+	items = [],
+	itemSelector = val => val as string,
+	loading,
+	loadingMessage = 'Loading...',
+	onItemSelect,
+	onValueChange,
+	value = '',
+	...otherProps
+}) => {
+	const activeListItemRef = useRef<HTMLLIElement>(null);
+	const inputRef = useRef<HTMLInputElement>(null);
+	const [activeIndex, setActiveIndex] = useState(-1);
+	const [active, setActive] = useState(!!value);
+
+	const keyHandler = useKeyHandlerForList({
+		activeListItemRef,
+		index: activeIndex,
+		inputRef,
+		onIndexChange: setActiveIndex,
+		onIndexSelect: (index: number) => onItemSelect(items[index]),
+		totalItems: items.length,
+	});
+
+	useEffect(() => {
+		setActive(!!value);
+	}, [value]);
+
+	useEffect(() => {
+		setActiveIndex(-1);
+	}, [active]);
+
+	return (
+		<ClayAutocomplete>
+			<ClayAutocomplete.Input
+				{...otherProps}
+				onChange={e => onValueChange(e.target.value)}
+				onKeyDown={keyHandler}
+				ref={inputRef}
+				value={value}
+			/>
+
+			<ClayAutocomplete.DropDown active={active} onSetActive={setActive}>
+				<ClayDropDown.ItemList>
+					{!loading &&
+						items.map((item, i) => (
+							<ClayAutocomplete.Item
+								active={i === activeIndex}
+								key={itemSelector(item)}
+								match={value.toString()}
+								onClick={() => onItemSelect(itemSelector(item))}
+								ref={
+									activeIndex === i
+										? activeListItemRef
+										: undefined
+								}
+								value={itemSelector(item)}
+							/>
+						))}
+
+					{loading && (
```

Comment:
I think maybe it would be better to add a new API to activate the message to cover the case that it only appears when the request exceeds 500ms.

I think we can add two checks here with this API, the default value of this new one is `null`, if `loading` is true it shows the message and if the other API is `true` or` false` just show the message accordingly new API.

Reply:
I went with the strict boolean since for this component itself I wanted to keep it straightforward for a simple use case. For use with a datasource, the consuming component would be the one to implement a 500ms delay. This is also the same reason the items prop is an array and not necessarily a datasource.

Reply:
Yes I agree with you I would like to keep our high-levels simple also so as not to become something complex and look like this https://github.com/liferay/clay/blob/2.x/packages/clay-autocomplete/src/ClayAutocomplete.js#L454-L795 🙂, but I think in this case, if we implement this loading here, we have to add the possibilities that people can do it, even if the data comes from outside for example.

Reply:
> if we implement this loading here, we have to add the possibilities that people can do it

Totally agree, but I think if we add the debounce and delay as an internal API for this component, its now assuming the consumer is using async data with a delay or its requiring them to pass a config to disable the delay. By not including it, its not strictly up to the consumer how they want to enable and use the loading flag, they can leverage it any way they would like since it is just a flag. Component's I've consumed in the past with internal delays or timers seem to always be tricky or require specific configuration to get it working how I'd want. From my perspective, I think its best to give users API that doesn't rely internal timers or delays, unless its an explicit async component.

We could think about creating a `AutoCompleteWithDataSource` type high-level component that would strictly be async.

Reply:
Oh no, I think you got it wrong, I do not want us to add some timer or delay, I understand your concern.

I still want us to keep the API `loading` and add one more, just to the loading of` loadingMessage` prop, so the dev will be able to decide if it will add the delay to the messages and activate the loading to the loadingIndicator as soon as it happens a new requisition.

Reply:
hmm, not sure I am tracking with you. Could you write up a code example API you are expecting? The `Loading...` message only appears if `loading={true}` and the input is focused and the dropdown is active.

Reply:
Np, this is an example of usage:

```jsx
<ClayInputWithAutocomplete
    loading={true}
    loadingMessage="Loading..."
    showLoadingMessage={true}
/>
```

in impl:

```jsx
export const ClayInputWithAutocomplete: React.FunctionComponent<IProps> = ({
    loading,
    loadingMessage = 'Loading...',
    showLoadingMessage = null, // is boolean or null
}) => {
    return (
        ...
            {((loading && !showLoadingMessage) || showLoadingMessage !== null) && (
                <ClayAutocomplete.Item value={loadingMessage} />
            )}
        ...
        ...
        {loading && <ClayAutocomplete.LoadingIndicator />}
    );
};
```

In the end I can just set `loading` to true and it will show the two feedback flags or I can control the two separately, set `loading` to `true` and `showLoadingMessage` to false and if pass the 500ms without resolving the request I set `showLoadingMessage` to true. Is that clearer?

Another question for the use of `<ClayAutocomplete.Item value={loadingMessage} />` the markup of the message is a bit different, you can check in v2 version. https://github.com/liferay/clay/blob/2.x/packages/clay-autocomplete/src/ClayAutocomplete.soy#L306-L311

Reply:
Gotcha, so overall you are just expecting an API to toggle the display of `loadingMessage`?

Reply:
Yeah